collect_app/src/main/java/org/odk/collect/android/fragments/ShowQRCodeFragment.java

```diff
@@ -0,0 +1,318 @@
+                // request was canceled...
+                ToastUtils.showShortToast("Scanning Cancelled");
+            } else {
+                applySettings(result.getContents());
+                return;
+            }
+        }
+
+        if (requestCode == SELECT_PHOTO) {
+            if (resultCode == Activity.RESULT_OK) {
+                try {
+                    final Uri imageUri = data.getData();
+                    final InputStream imageStream = getActivity().getContentResolver()
+                            .openInputStream(imageUri);
+
+                    final Bitmap selectedImage = BitmapFactory.decodeStream(imageStream);
+                    scanQRImage(selectedImage);
+                } catch (FileNotFoundException e) {
+                    Timber.e(e);
+                }
+            } else {
+                ToastUtils.showShortToast("Cancelled");
+            }
+        }
+    }
+
+    private void scanQRImage(Bitmap bitmap) {
+        int[] intArray = new int[bitmap.getWidth() * bitmap.getHeight()];
+
+        //copy pixel data from bitmap into the array
+        bitmap.getPixels(intArray, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());
+
+        LuminanceSource source = new RGBLuminanceSource(bitmap.getWidth(), bitmap.getHeight(), intArray);
+        BinaryBitmap binaryBitmap = new BinaryBitmap(new HybridBinarizer(source));
+
+        Reader reader = new QRCodeMultiReader();
+        try {
+            Result result = reader.decode(binaryBitmap);
+            applySettings(result.getText());
+        } catch (FormatException | NotFoundException | ChecksumException e) {
+            Timber.i(e);
+            ToastUtils.showLongToast("QR Code not found in the selected image");
+        }
+    }
+
+    private void applySettings(String content) {
+        String decompressedData;
+        try {
+            decompressedData = TextUtils.decompress(content);
+        } catch (DataFormatException e) {
+            Timber.e(e);
+            ToastUtils.showShortToast("QR Code does not contains valid settings");
+            return;
+        } catch (IOException e) {
+            Timber.e(e);
+            return;
+        }
+
+        JSONObject jsonObject;
+        try {
+            jsonObject = new JSONObject(decompressedData);
+            SharedPreferences.Editor editor = settings.edit();
+            editor.putString(KEY_PROTOCOL, jsonObject.getString(KEY_PROTOCOL));
+            editor.putString(KEY_SERVER_URL, jsonObject.getString(KEY_SERVER_URL));
+            editor.putString(KEY_GOOGLE_SHEETS_URL, jsonObject.getString(KEY_GOOGLE_SHEETS_URL));
+            editor.putString(KEY_FORMLIST_URL, jsonObject.getString(KEY_FORMLIST_URL));
+            editor.putString(KEY_SUBMISSION_URL, jsonObject.getString(KEY_SUBMISSION_URL));
+            editor.putString(KEY_USERNAME, jsonObject.getString(KEY_USERNAME));
+            editor.putString(KEY_PASSWORD, jsonObject.getString(KEY_PASSWORD));
```

Comment:
Can these be in a loop that iterates over an array of the relevant KEY constants?